src/nspawn/nspawn-network.c

```diff
@@ -541,6 +543,173 @@ static int needs_rename(sd_netlink **rtnl, sd_device *dev, const char *name) {
         return !strv_contains(altnames, name);
 }
 
+typedef struct DeviceMonitorData {
+        int ifindex;
+        const char *name;
+        sd_device *dev;
+} DeviceMonitorData;
+
+static void device_monitor_data_done(DeviceMonitorData *data) {
+        if (!data)
+                return;
+
+        sd_device_unref(data->dev);
+}
+
+static int device_monitor_handler(sd_device_monitor *monitor, sd_device *device, void *userdata) {
+        DeviceMonitorData *data = ASSERT_PTR(userdata);
+
+        assert(device);
+        assert(data->ifindex > 0);
+        assert(data->name);
+        assert(!data->dev);
+
+        if (device_for_action(device, SD_DEVICE_REMOVE))
+                return 0;
+
+        int ifindex;
+        if (sd_device_get_ifindex(device, &ifindex) < 0 || ifindex != data->ifindex)
+                return 0;
+
+        const char *name;
+        if (sd_device_get_sysname(device, &name) < 0 || !streq(name, data->name))
+                return 0;
+
+        if (device_is_renaming(device))
+                return 0;
+
+        data->dev = sd_device_ref(device);
+        return sd_event_exit(sd_device_monitor_get_event(monitor), 0);
+}
+
+static int rename_and_wait(sd_netlink **rtnl, sd_device *dev, const char *name, sd_device **ret_dev) {
+        _cleanup_(sd_device_monitor_unrefp) sd_device_monitor *monitor = NULL;
+        _cleanup_(sd_event_unrefp) sd_event *event = NULL;
+        int r;
+
+        assert(rtnl);
+        assert(dev);
+        assert(name);
+        assert(ret_dev);
+
+        /* The command NL80211_CMD_SET_WIPHY_NETNS below passes phy instead of network interface,
+         * and does not take an interface name in the passed network namespace. Hence, we need to
+         * rename the interface before passing phy to the netns. */
```

Comment:
so, if another interface already exists under the target name in the source namespace then the whole operation will fail, no? That would suck, no?

Maybe create a throw-away middle netns? i.e. open a new netns, migrate the iface there, then rename it in there, and then migrate it to the target ns? that way the interface will only exist under the old name in the source netns, and only under the new name in the target netns.

or am i missing something here, and this is not an issue?

Reply:
Sounds nice. I will try.